lib/ua_update_messaging.py

```diff
@@ -0,0 +1,317 @@
+            grace_period_remaining = (
+                defaults.CONTRACT_EXPIRY_GRACE_PERIOD_DAYS + remaining_days
+            )
+            pkgs_msg = MESSAGE_CONTRACT_EXPIRED_GRACE_PERIOD_TMPL.format(
+                expired_date=cfg.contract_expiry_datetime.strftime("%d %b %Y"),
+                remaining_days=grace_period_remaining,
+                url=defaults.BASE_UA_URL,
+            )
+            # Same cautionary message when in grace period
+            motd_pkgs_msg = motd_no_pkgs_msg = no_pkgs_msg = pkgs_msg
+        elif expiry_status == ContractExpiryStatus.EXPIRED:
+            series = util.get_platform_info()["series"]
+            if ent.name == "esm-apps" and series in ("trusty", "xenial"):
+                no_warranty = MESSAGE_UBUNTU_NO_WARRANTY
+            else:
+                no_warranty = ""
+            pkgs_msg = (
+                MESSAGE_CONTRACT_EXPIRED_APT_PKGS_TMPL.format(
+                    pkg_num=tmpl_pkg_count_var,
+                    pkg_names=tmpl_pkg_names_var,
+                    title=ent.title,
+                    url=defaults.BASE_UA_URL,
+                )
+                + no_warranty
+            )
+            no_pkgs_msg = (
+                MESSAGE_CONTRACT_EXPIRED_APT_NO_PKGS_TMPL.format(
+                    title=ent.title, url=defaults.BASE_ESM_URL
+                )
+                + no_warranty
+            )
+            motd_no_pkgs_msg = no_pkgs_msg
+            motd_pkgs_msg = (
+                MESSAGE_CONTRACT_EXPIRED_MOTD_PKGS_TMPL.format(
+                    title=ent.title,
+                    pkg_num=tmpl_pkg_count_var,
+                    url=defaults.BASE_ESM_URL,
+                )
+                + no_warranty
+            )
+    else:  # Apps not enabled on ESM Ubuntu release
+        pkgs_msg = MESSAGE_DISABLED_APT_PKGS_TMPL.format(
+            title=ent.title,
+            pkg_num=tmpl_pkg_count_var,
+            pkg_names=tmpl_pkg_names_var,
+            url=defaults.BASE_ESM_URL,
+        )
+        no_pkgs_msg = MESSAGE_DISABLED_MOTD_NO_PKGS_TMPL.format(
+            title=ent.title, url=defaults.BASE_ESM_URL
+        )
+        motd_pkgs_msg = MESSAGE_DISABLED_MOTD_PKGS_TMPL.format(
+            title=ent.title,
+            pkg_num=tmpl_pkg_count_var,
+            url=defaults.BASE_ESM_URL,
+        )
+        motd_no_pkgs_msg = MESSAGE_DISABLED_MOTD_NO_PKGS_TMPL.format(
+            title=ent.title, url=defaults.BASE_ESM_URL
+        )
+
+    msg_dir = os.path.join(cfg.data_dir, "messages")
+    _write_template_or_remove(no_pkgs_msg, os.path.join(msg_dir, no_pkgs_file))
+    _write_template_or_remove(pkgs_msg, os.path.join(msg_dir, pkgs_file))
+    _write_template_or_remove(
+        motd_no_pkgs_msg, os.path.join(msg_dir, motd_no_pkgs_file)
+    )
+    _write_template_or_remove(
+        motd_pkgs_msg, os.path.join(msg_dir, motd_pkgs_file)
+    )
+
+
+def write_apt_pre_invoke_templates(cfg: config.UAConfig) -> None:
```

Comment:
Consider renaming this function since this does more than just apt_pre_invoke messages